sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala

```diff
@@ -275,10 +272,19 @@ case class CachedRDDBuilder(
         storageLevel,
         cachedPlan.conf)
     }
-    val cached = cb.map { batch =>
-      sizeInBytesStats.add(batch.sizeInBytes)
-      rowCountStats.add(batch.numRows)
-      batch
+    val cached = cb.mapPartitionsInternal { it =>
+      new Iterator[CachedBatch] {
+        TaskContext.get().addTaskCompletionListener[Unit](_ => {
```

Comment:
we can register this listener before returning the wrapping iterator.

Reply:
oh, somehow I put the code here...